File: python/linked_list/234. Palindrome Linked List.py
'''
Two pointer I:
Time: O(n)
Space: O(n)
Two pointer II:
Time: O(n)
Space: O(1)
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # Two pointer I
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        nums = []
        # linked list to array
        while head:
            nums.append(head.val)
            head = head.next

        l, r = 0, len(nums) - 1

        while l <= r:
            if nums[l] != nums[r]:
                return False
            l += 1
            r -= 1
        
        return True
    
    # Two pointer I
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # Initialize two pointers
        fast, slow = head, head

        # move slow to the mid
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        
        # Reverse second half
        prev = None
        while slow:
            temp = slow.next
            slow.next = prev
            prev = slow
            slow = temp

        # Check palindrome
        l, r = head, prev
        while r:  # mid point is None
            if l.val != r.val:
                return False
            l = l.next
            r = r.next

        return True

File: javascript/linked_list/234. Palindrome Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
// Two pointer I
var isPalindrome = function(head) {
    const nums = [];

    while (head) {
        nums.push(head.val);
        head = head.next;
    }
    let l = 0, r = nums.length - 1;
    while (l <= r) {
        if (nums[l] != nums[r]) return false;

        l++;
        r--;
    }
    return true;
};
// Two pointer I
var isPalindrome = function(head) {
    let fast = head, slow = head;

    // mid
    while (fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
    };
    // reverse
    let prev = null;
    while (slow) {
        let temp = slow.next;
        slow.next = prev;
        prev = slow;
        slow = temp;
    };
    // check
    let l = head, r = prev;
    while (r) {
        if (l.val != r.val) return false;

        l = l.next;
        r = r.next;
    };
    return true;
};